### PR TITLE
`azurerm_container_group` - Support of `dns_config`

### LINKED ISSUE #6360: Support for azurerm_container_group for setting DNS in private networks
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

https://github.com/MicrosoftFeedback/aci-issues/issues/25

It's possible for azure container instances to support DNS inside a custom VNet, but this isn't exposed in the `azurerm_container_group` options. Please add the ability to support this. My use case is that I have a virtual network in which I want the ACI that is running to be able to use services by DNS rather than IP. Specifically I'm trying to setup a container that runs a bridge to with Azure Relay, and want the relay listener running in ACI.

### New or Affected Resource(s)

* azurerm_container_group

### Potential Terraform Configuration

```hcl
resource "azurerm_container_group" "hybrid-listener" {
  name                = "aci-listener"
  location            = "West Europe"
  resource_group_name = azurerm_resource_group.azwe01rg01_t_relay.name
  ip_address_type     = "Private"
  network_profile_id  = azurerm_network_profile.azwe01acr01_profile.id
  os_type             = "Linux"
                                                   1                                             2
  container {
    name   = "azbridge"
    image  = "azbridge:latest"
    cpu    = "0.5"
    memory = "1.5"

    ports {
      port     = 9998
      protocol = "UDP"
    }
  }

  // See: https://github.com/Azure/azure-rest-api-specs/blob/21f0dcbb89544101d79be7fc889a86ce0911901b/specification/containerinstance/resource-manager/Microsoft.ContainerInstance/stable/2018-10-01/examples/ContainerGroupsCreateOrUpdate.json#L73
  dns_config {
    name_servers = ["10.0.1.1", "10.0.1.2"]
    search_domains = "internal.mydomain"
    options = "ndots:2"
  }
}
```

### References

* Found when trying to test and work around #1697

### PR DESCRIPTION
**Pull request by zparnold**
Closes #6360 

This PR attempts to add in the DNS config functionality for ACI. Please let me know if I missed something for the AzureRM provider. I think? I hit all the required terraform bits. Reference to the JSON schema for ACI below that I implemented.

https://github.com/Azure/azure-rest-api-specs/blob/21f0dcbb89544101d79be7fc889a86ce0911901b/specification/containerinstance/resource-manager/Microsoft.ContainerInstance/stable/2018-10-01/examples/ContainerGroupsCreateOrUpdate.json#L73

### CODE REVIEW

**Comment by zparnold** [comment]
@jackofallops I'm not sure if you're a maintainer, but if you are, and aren't too busy could I please have a review?

**Comment by zparnold** [comment]
Bump

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:430]: Can we change the type into `schema.TypeSet` as azure API doesn't always guarantee the order to be same as in request and in response when later get.

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:438]: Can we change the type to be `schema.TypeSet` here?

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:444]: Does `options` allow multiple values (e.g. separated by space)? If so, can we also change it into `schema.TypeSet` here?

Besides, shall we mark it as `Optional` as it is an option itself?

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:1419]: If the field is absent in `input` here, can we assign a zero value of that type into the map as other resources' flatten function does?

(same applies to other properties below)

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:1451]: Can we simply return the structure literal here:

```suggestion
	return &containerinstance.DNSConfiguration{
		Options:       utils.String(config["options"].(string)),
		SearchDomains: utils.String(config["search_domains"].(string)),
		NameServers:   &ns,
	}
```

**Review by magodo — CHANGES_REQUESTED** [review]
@zparnold Thank you for this PR :+1:

Having look thourgh the code and left some comments mainly on the schema design. Besides, can you please add the test for this new property to ensure it works?

Thank you!

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go:444]: I think it does I was just trying to stay as close as possible to the AzureRM go client impl, will change :)
```go
// DNSConfiguration DNS configuration for the container group.
type DNSConfiguration struct {
	// NameServers - The DNS servers for the container group.
	NameServers *[]string `json:"nameServers,omitempty"`
	// SearchDomains - The DNS search domains for hostname lookup in the container group.
	SearchDomains *string `json:"searchDomains,omitempty"`
	// Options - The DNS options for the container group.
	Options *string `json:"options,omitempty"`
}

```

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go:438]: I think it does I was just trying to stay as close as possible to the AzureRM go client impl, would you like me to change it? Below is the Azure Go Client struct
```go
// DNSConfiguration DNS configuration for the container group.
type DNSConfiguration struct {
	// NameServers - The DNS servers for the container group.
	NameServers *[]string `json:"nameServers,omitempty"`
	// SearchDomains - The DNS search domains for hostname lookup in the container group.
	SearchDomains *string `json:"searchDomains,omitempty"`
	// Options - The DNS options for the container group.
	Options *string `json:"options,omitempty"`
}

```

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go:1419]: I'm assuming, but could be wrong here that if a pointer to a string that is not nil will be initialized to its zero value automatically. So I think by putting `output["nameservers"] = *v` when I dereference the non-nil pointer I _at least_ get `""` and at most I get some non-empty value. Am I correct? I'm still learning go :)

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [reply azurerm/internal/services/containers/container_group_resource.go:438]: Yeah, it is weired why Swagger defines that field as a `string` as it is actually a space delimited list. I prefer using a `Set` here to serve better user experience.

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [reply azurerm/internal/services/containers/container_group_resource.go:444]: Yeah, it is weired why Swagger defines that field as a string as it is actually a space delimited list. I prefer using a Set here to serve better user experience.

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [reply azurerm/internal/services/containers/container_group_resource.go:1417]: You're right. While what I suggested is to do something like:

```go
searchDomains := ""
if input.SearchDomains != nil {
    searchDomains = *input.SearchDomains
}
output["search_domains"] = searchDomains
```

The intent is to always set `search_domains`.

**Comment by ash007osh** [comment]
Has this been implemented in Terraform release?

**Comment by zparnold** [comment]
@magodo I hopefully have addressed all PR comments.

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go]: Please check I believe I have resolved

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go]: I changed them both to set types

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go]: I changed them all to sets

File: website/docs/r/container_group.html.markdown
Comment (magodo) [inline website/docs/r/container_group.html.markdown:249]:
```suggestion
* `nameservers` - (Required) A list of nameservers the containers will search out to resolve requests.
* `search_domains` - (Required) A list of search domains that DNS requests will search along.
* `options` - (Required) A list of [resolver configuration options](https://man7.org/linux/man-pages/man5/resolv.conf.5.html).
```

Can we modify the wordings and formattings a bit?

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:1448]: Can we change the function signature a bit to accept a `[]interface{}` instead, and modify the first `if` logic a bit as [this example expand function](https://github.com/terraform-providers/terraform-provider-azurerm/blob/aad52564c4fad1197ed4ab03f62e35d0ffc0f63b/azurerm/internal/services/containers/container_group_resource.go#L876)?

(since the type of `dns_config` is always guaranteed to be `schema.Set` in this case)

File: azurerm/internal/services/containers/container_group_resource.go
Comment (magodo) [inline azurerm/internal/services/containers/container_group_resource.go:1417]: Can we remove this local variable here so as to be consistent with other flatten functions? ([example](https://github.com/terraform-providers/terraform-provider-azurerm/blob/aad52564c4fad1197ed4ab03f62e35d0ffc0f63b/azurerm/internal/services/containers/container_group_resource.go#L979))

**Review by magodo — CHANGES_REQUESTED** [review]
Thank you for the update! I have left some additional mostly minor comments that once addressed this should be good to merge 👍

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go]: This seems a bit more than minor, and the example you have provided was a TypeList not a TypeSet. I will attempt to refactor for this, but when I was testing last night it broke...a lot.

File: azurerm/internal/services/containers/container_group_resource.go
Comment (zparnold) [reply azurerm/internal/services/containers/container_group_resource.go]: Sure.

**Comment by zparnold** [comment]
@magodo anything else wrong with this?

**Comment by zparnold** [comment]
@magodo when I take your suggestion and do the `[]interface{}` approach here is what I get when I run acceptance tests locally. I strongly advise that we go back to my original function signature and design:

```
=== RUN   TestAccAzureRMContainerGroup_virtualNetwork
=== PAUSE TestAccAzureRMContainerGroup_virtualNetwork
=== CONT  TestAccAzureRMContainerGroup_virtualNetwork
panic: interface conversion: interface {} is *schema.Set, not []interface {}

goroutine 495 [running]:
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers.expandContainerGroupDnsConfig(0x55f5760, 0xc001243320, 0xc001338f80)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers/container_group_resource.go:1447 +0x7cf
github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers.resourceArmContainerGroupCreate(0xc000b8dce0, 0x51bb4a0, 0xc000824dc0, 0x0, 0x0)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers/container_group_resource.go:508 +0x4ff
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).Apply(0xc00050d5f0, 0xc00207d5e0, 0xc000b37cc0, 0x51bb4a0, 0xc000824dc0, 0xc0010d9901, 0xc001409770, 0xc0010d99a0)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/resource.go:310 +0x365
github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Provider).Apply(0xc000132b00, 0xc001db79c8, 0xc00207d5e0, 0xc000b37cc0, 0xc0010d9738, 0xc0012db9f0, 0x5187860)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/helper/schema/provider.go:294 +0x99
github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ApplyResourceChange(0xc0012dbb28, 0x623d6c0, 0xc0013a8030, 0xc000b8d7a0, 0xc0012dbb28, 0xc0013a8030, 0xc000329b30)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin/grpc_provider.go:885 +0x8b4
github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ApplyResourceChange_Handler(0x56112e0, 0xc0012dbb28, 0x623d6c0, 0xc0013a8030, 0xc000e03440, 0x0, 0x623d6c0, 0xc0013a8030, 0xc0018bb800, 0x79c)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5/tfplugin5.pb.go:3305 +0x217
google.golang.org/grpc.(*Server).processUnaryRPC(0xc0016d7680, 0x62a48e0, 0xc0017ae180, 0xc000e78b00, 0xc0017471d0, 0x94dc0c0, 0x0, 0x0, 0x0)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1024 +0x4f4
google.golang.org/grpc.(*Server).handleStream(0xc0016d7680, 0x62a48e0, 0xc0017ae180, 0xc000e78b00, 0x0)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:1313 +0xd97
google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0016375e0, 0xc0016d7680, 0x62a48e0, 0xc0017ae180, 0xc000e78b00)
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:722 +0xbb
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/Users/zarnold/go/src/github.com/terraform-providers/terraform-provider-azurerm/vendor/google.golang.org/grpc/server.go:720 +0xa1

Process finished with exit code 1
```

**Comment by zparnold** [comment]
I have added something which fixes the acceptance test. Let me know if it breaks something else.

**Comment by magodo** [comment]
LGTM!

image

**Pull request merged**
